cache/cache.go

```diff
@@ -1,95 +1,258 @@
+		return atime.Get(files[i].info).Before(atime.Get(files[j].info))
+	})
+
+	for _, f := range files {
+		key := f.name[len(c.dir)+1:]
+		c.lru.Add(key, &lruItem{
+			size:      f.info.Size(),
+			committed: true,
+		})
+	}
 }
 
-func (c *cache) AddFile(hash string, size int64) {
+func (c *fsCache) Put(key string, size int64, expectedSha256 string, r io.Reader) (err error) {
 	c.mux.Lock()
-	defer c.mux.Unlock()
-	_, ok := c.files[hash]
+
+	// If `key` is already in the LRU, do nothing. This applied to both
+	// committed an uncommitted files (we don't want to upload again if an upload
+	// of the same file is already in progress).
+	if _, exists := c.lru.Get(key); exists {
+		c.mux.Unlock()
+		return
+	}
+
+	// Try to add the item to the LRU.
+	newItem := &lruItem{
+		size:      size,
+		committed: false,
+	}
+	ok := c.lru.Add(key, newItem)
+	c.mux.Unlock()
 	if !ok {
-		c.files[hash] = size
-		c.currSize += size
+		return &ErrTooBig{}
 	}
+
+	// By the time this function exits, we should either mark the LRU item as committed
+	// (if the upload went well), or delete it. Capturing the flag variable is not very nice,
+	// but this stuff is really easy to get wrong without defer().
+	shouldCommit := false
+	defer func() {
+		c.mux.Lock()
+		defer c.mux.Unlock()
+
+		if shouldCommit {
+			newItem.committed = true
+		} else {
+			c.lru.Remove(key)
+		}
+	}()
+
+	// Download to a temporary file
+	f, err := ioutil.TempFile(c.dir, "upload")
+	if err != nil {
+		return
+	}
+	defer os.Remove(f.Name())
+
+	if expectedSha256 != "" {
+		hasher := sha256.New()
+		if _, err = io.Copy(io.MultiWriter(f, hasher), r); err != nil {
+			return
+		}
+		actualHash := hex.EncodeToString(hasher.Sum(nil))
+		if actualHash != expectedSha256 {
+			err = errors.New(fmt.Sprintf(
+				"hashsums don't match. Expected %s, found %s", expectedSha256, actualHash))
+			return
+		}
+	} else {
+		if _, err = io.Copy(f, r); err != nil {
+			return
+		}
+	}
+
+	if err := f.Sync(); err != nil {
+		return err
+	}
+
+	if err := f.Close(); err != nil {
+		return err
+	}
+
+	// Rename to the final path
+	blobPath := c.pathForKey(key)
+	err = os.Rename(f.Name(), blobPath)
+	// Only commit if renaming succeeded
+	if err == nil {
+		// This flag is used by the defer() block above.
+		shouldCommit = true
+	}
+
+	return
 }
 
-func (c *cache) RemoveFile(hash string) int64 {
-	c.mux.Lock()
-	defer c.mux.Unlock()
-	size := c.files[hash]
-	c.currSize -= size
-	delete(c.files, hash)
-	return size
+func (c *fsCache) Get(key string, w http.ResponseWriter) (ok bool, err error) {
+	ok = func() bool {
+		c.mux.Lock()
+		defer c.mux.Unlock()
+
+		val, found := c.lru.Get(key)
+		// Uncommitted (i.e. uploading items) should be reported as not ok
+		return found && val.(*lruItem).committed
+	}()
+
+	if !ok {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	blobPath := c.pathForKey(key)
+	f, err := os.Open(blobPath)
+	if err != nil {
+		return
+	}
+	defer f.Close()
+
+	_, err = io.Copy(w, f)
```

Comment:
I think it's easiest if you just use `http.ServeFile` here as it also sets all the necessary headers. This is also why the `Content Length` header is not set and you had to uncomment the testing code.

Reply:
I added the missing header back. I'd rather not use `serveFile` since it couples the cache to the http request and has a bunch of magic for autodetecting types, etc..